File: backend/api-gateway/internal/models/graph.go
package models

// NodeType represents the type of a graph node.
type NodeType string

const (
	NodeTypeAsset         NodeType = "asset"
	NodeTypeVulnerability NodeType = "vulnerability"
	NodeTypeIdentity      NodeType = "identity"
	NodeTypePermission    NodeType = "permission"
	NodeTypeNetwork       NodeType = "network"
	NodeTypeProcess       NodeType = "process"
)

// EdgeType represents the type of a graph edge (relationship).
type EdgeType string

const (
	EdgeTypeConnectsTo    EdgeType = "connects_to"
	EdgeTypeHasAccess     EdgeType = "has_access"
	EdgeTypeExploits      EdgeType = "exploits"
	EdgeTypeDependsOn     EdgeType = "depends_on"
	EdgeTypeContains      EdgeType = "contains"
	EdgeTypeTrustsTo      EdgeType = "trusts_to"
	EdgeTypeAuthenticates EdgeType = "authenticates"
)

// GraphNode represents a node in the attack graph.
type GraphNode struct {
	ID         string            `json:"id"`
	Type       NodeType          `json:"type"`
	Label      string            `json:"label"`
	Properties map[string]string `json:"properties,omitempty"`
	RiskScore  float64           `json:"risk_score"`
}

// GraphEdge represents an edge (relationship) in the attack graph.
type GraphEdge struct {
	ID         string            `json:"id"`
	Source     string            `json:"source"`
	Target     string            `json:"target"`
	Type       EdgeType          `json:"type"`
	Weight     float64           `json:"weight"`
	Properties map[string]string `json:"properties,omitempty"`
}

// GraphQuery represents parameters for querying the graph.
type GraphQuery struct {
	NodeTypes  []NodeType `json:"node_types,omitempty"`
	EdgeTypes  []EdgeType `json:"edge_types,omitempty"`
	Limit      int        `json:"limit,omitempty"`
	Offset     int        `json:"offset,omitempty"`
	MinRisk    float64    `json:"min_risk,omitempty"`
	MaxRisk    float64    `json:"max_risk,omitempty"`
	SearchTerm string     `json:"search_term,omitempty"`
}

// PathResult represents a path between two nodes.
type PathResult struct {
	Nodes     []GraphNode `json:"nodes"`
	Edges     []GraphEdge `json:"edges"`
	TotalRisk float64     `json:"total_risk"`
	Length    int         `json:"length"`
}

// BlastRadiusResult represents the blast radius from a given node.
type BlastRadiusResult struct {
	OriginNode     GraphNode   `json:"origin_node"`
	AffectedNodes  []GraphNode `json:"affected_nodes"`
	AffectedEdges  []GraphEdge `json:"affected_edges"`
	TotalImpact    float64     `json:"total_impact"`
	Depth          int         `json:"depth"`
}

// DependencyResult represents the dependencies of a given node.
type DependencyResult struct {
	Node         GraphNode   `json:"node"`
	Dependencies []GraphNode `json:"dependencies"`
	Dependents   []GraphNode `json:"dependents"`
}

File: shared/types/src/graph.ts
/**
 * Graph engine types for the ARES-X platform.
 * @module graph
 */

/** Type classification for graph nodes. */
export enum NodeType {
  Server = 'server',
  Workstation = 'workstation',
  NetworkDevice = 'network_device',
  Database = 'database',
  Application = 'application',
  User = 'user',
  ServiceAccount = 'service_account',
  CloudResource = 'cloud_resource',
  Container = 'container',
  Identity = 'identity',
}

/** Type classification for graph edges (relationships). */
export enum EdgeType {
  NetworkAccess = 'network_access',
  CredentialAccess = 'credential_access',
  PrivilegeEscalation = 'privilege_escalation',
  LateralMovement = 'lateral_movement',
  DataFlow = 'data_flow',
  TrustRelationship = 'trust_relationship',
  Membership = 'membership',
  Dependency = 'dependency',
}

/** Represents a node (vertex) in the attack graph. */
export interface GraphNode {
  /** Unique node identifier. */
  id: string;
  /** Display name for the node. */
  name: string;
  /** Node type classification. */
  type: NodeType;
  /** Key-value properties associated with the node. */
  properties: Record<string, string>;
  /** Computed risk score (0.0 - 10.0). */
  riskScore: number;
  /** Current status of the node. */
  status: 'active' | 'inactive' | 'compromised' | 'unknown';
  /** Timestamp when the node was created. */
  createdAt: string;
  /** Timestamp of last update. */
  updatedAt: string;
}

/** Represents a directed edge (relationship) between two nodes. */
export interface GraphEdge {
  /** Unique edge identifier. */
  id: string;
  /** Source node identifier. */
  sourceId: string;
  /** Target node identifier. */
  targetId: string;
  /** Relationship type. */
  type: EdgeType;
  /** Edge weight representing traversal cost/difficulty. */
  weight: number;
  /** Key-value properties associated with the edge. */
  properties: Record<string, string>;
  /** Timestamp when the edge was created. */
  createdAt: string;
}

/** Query parameters for fetching graph data. */
export interface GraphQuery {
  /** Filter by node types. */
  nodeTypes?: NodeType[];
  /** Filter by edge types. */
  edgeTypes?: EdgeType[];
  /** Custom property filters. */
  filters?: Record<string, string>;
  /** Minimum risk score threshold. */
  minRiskScore?: number;
  /** Maximum depth for traversal queries. */
  maxDepth?: number;
  /** Maximum number of results. */
  limit?: number;
}

/** Response containing graph data. */
export interface GraphResponse {
  /** Returned nodes. */
  nodes: GraphNode[];
  /** Returned edges. */
  edges: GraphEdge[];
  /** Total count of matching nodes. */
  totalNodes: number;
  /** Total count of matching edges. */
  totalEdges: number;
}

/** Result of a path-finding operation. */
export interface PathResult {
  /** Ordered list of nodes in the path. */
  nodes: GraphNode[];
  /** Ordered list of edges in the path. */
  edges: GraphEdge[];
  /** Total weight (cost) of the path. */
  totalWeight: number;
  /** Number of hops in the path. */
  hopCount: number;
}

/** Result of a blast radius computation. */
export interface BlastRadius {
  /** The origin node from which blast radius was computed. */
  originId: string;
  /** Nodes affected within the blast radius. */
  affectedNodes: GraphNode[];
  /** Edges traversed during computation. */
  traversedEdges: GraphEdge[];
  /** Total number of affected nodes. */
  totalAffected: number;
  /** Aggregate risk score across all affected nodes. */
  aggregateRiskScore: number;
  /** Count of affected nodes by type. */
  affectedByType: Record<string, number>;
}
